app/services/scraper/bedelias.rb

```diff
@@ -0,0 +1,141 @@
+
+    prerequisites.each do |prerequisite_tree|
+      add_missing_exams_and_subjects(prerequisite_tree, subjects)
+    end
+
+    File.write(Rails.root.join("db/data/scraped_subject_groups.yml"), groups.deep_stringify_keys.to_yaml)
+    File.write(Rails.root.join("db/data/scraped_subjects.yml"), subjects.deep_stringify_keys.to_yaml)
+    File.write(Rails.root.join("db/data/scraped_prerequisites.yml"), { prerequisites: }.deep_stringify_keys.to_yaml)
+  rescue
+    Rails.logger.info save_screenshot
+    raise
+  end
+
+  private
+
+  def go_to_groups_and_subjects_page
+    visit "https://bedelias.udelar.edu.uy"
+    click_on "PLANES DE ESTUDIO"
+    click_on "Planes de estudio / Previas"
+
+    execute_script("$.fx.off = true;") # Disable jQuery effects
+
+    find('h3', text: 'TECNOLOGÍA Y CIENCIAS DE LA NATURALEZA').click
+    find('span', text: 'FING - FACULTAD DE INGENIERÍA', visible: false).click
+
+    within('tr', text: 'INGENIERIA EN COMPUTACION', match: :prefer_exact) do
+      find('.ui-row-toggler').click
+    end
+    within('.ui-expanded-row-content', text: 'Planes') do
+      find('tr', text: '1997').click_on "Ver más datos"
+    end
+  end
+
+  def process_groups_and_subjects(groups, subjects)
+    node_type = "[data-nodetype='Grupo']"
+    selector = ".arbolComposicion #{node_type}:not(:has(#{node_type})) > .ui-treenode-content > .ui-treenode-label"
+
+    all(selector, visible: false).each do |node|
+      group_code, name, credits = CODE_NAME_CREDITS_REGEX.match(node.text(:all)).captures
+      groups[group_code] = { code: group_code, name:, min_credits: credits.to_i }
+
+      node.all(:xpath, '..//..//li[@data-nodetype="Materia"]/span', visible: false).each do |subject_node|
+        code, name, credits = CODE_NAME_CREDITS_REGEX.match(subject_node.text(:all)).captures
+        subjects[code] = { code:, name:, credits: credits.to_i, has_exam: false, subject_group: group_code }
+      end
+    end
+  end
+
+  def go_to_prerequisites_page
+    click_on 'Sistema de previaturas'
+  end
+
+  def process_prerequisites(subjects)
+    total_pages = /\(\d+ de (\d+)\)/.match(find(".ui-paginator-current").text).captures[0]
+
+    1.upto(MAX_PAGES || total_pages.to_i).flat_map do |page|
+      Rails.logger.info "Page #{page}"
+
+      go_to_page(page, total_pages)
+
+      all('[data-ri]').map { |node| node['data-ri'] }.map do |row_id|
+        go_to_page(page, total_pages)
+
+        row = find("[data-ri='#{row_id}']")
+        name, type = row.all('td')[0..1]
+        Rails.logger.info "#{name.text} (#{type.text})"
+
+        subject_code = /\A(\w+) - .*\z/.match(name.text).captures[0]
+        is_exam = type.text == "Examen"
+
+        subjects[subject_code][:has_exam] ||= is_exam if subjects[subject_code]
+
+        row.click_on "Ver más"
+
+        prereq = Scraper::PrerequisitesTree.prerequisite_tree(find("[data-rowkey='root']"))
+
+        click_on 'Volver'
+
+        prereq.merge(subject_code:, is_exam:)
+      end
```

Comment:
we can't make it exactly like this because each time we visit the details for an approvable and go back to the index page, the nodes are different. That's why we need to get the row ids first, and then find each particular row.

I added a comment explaining this and did the other changes (moved the go_to_page and renamed the variable) in 0f20472d32f13efe8615184cd5bc4c96ddcba4a5

Reply:
Yeah good call, I always forget about that 😅